packages/components/src/components/files/FileTree.tsx

```diff
@@ -0,0 +1,359 @@
+                    margin: '-7px 0 0 -7px',
+                    height: '14px'
+                },
+                height: 14,
+                width: 14,
+                arrow: {
+                    fill: '#777',
+                    strokeWidth: 0
+                }
+            },
+            header: {
+                base: {
+                    display: 'inline-block',
+                    verticalAlign: 'top',
+                    color: '#777'
+                },
+                connector: {
+                    width: '2px',
+                    height: '12px',
+                    borderLeft: 'solid 2px black',
+                    borderBottom: 'solid 2px black',
+                    position: 'absolute',
+                    top: '0px',
+                    left: '-21px'
+                },
+                title: {
+                    lineHeight: '24px',
+                    verticalAlign: 'middle'
+                }
+            },
+            subtree: {
+                listStyle: 'none',
+                paddingLeft: '19px'
+            },
+            loading: {
+                color: '#777'
+            }
+        }
+    }
+};
+
+interface FileTreeProps {
+    loadData: any,
+    onFileSelect?: (name: string, path: string, checked: boolean, isDirectory: boolean) => any
+}
+
+interface FileTreeState {
+    cursor: any,
+    checked: List<string>
+    data: any
+    loading: boolean
+    error?: string
+}
+
+export class FileTree extends React.Component<FileTreeProps, FileTreeState> {
+
+    constructor(props) {
+        super(props);
+
+        this.state = {
+            cursor: undefined,
+            checked: List<string>(),
+            data: [],
+            loading: false,
+            error: undefined
+        }
+    }
+
+    CHECK_ID_PREFIX = 'filetree-check-';
+    DEFAULT_ROOT_PREFIX = '|root';
+    EMPTY_FILE_NAME = '*empty';
+
+    componentDidMount(): void {
+        const { loadData } = this.props;
+        const { loading } = this.state;
+
+        this.setState((state) => ({loading: true}));
+        loadData().then((data) => {
+            let loadedData = data;
+
+            // treebeard has bugs when there is not a single root node
+            if (Array.isArray(data)) {
+
+                if (data.length < 1) {
+                    data = [{id: this.DEFAULT_ROOT_PREFIX + "|" + this.EMPTY_FILE_NAME, active: false, name: "empty"}]
+                }
+
+                loadedData = {
+                    name: 'root',
+                    id: this.DEFAULT_ROOT_PREFIX,  // Special id
+                    children: data,
+                    toggled: true
+                }
+            }
+
+            if (!loadedData.id) {
+                loadedData.id = loadedData.name;
+            }
+
+            this.setState(() => ({data: loadedData, loading: false}))
+        }).catch((reason: any) => {
+            this.setState(() => ({error: reason}));
+        })
+    }
+
+    Header = (props) => {
```

Comment:
I don't think it should be defined on the class. It can be  a separate component defined in the same file, there's nothing wrong with that. It's really weird to attach it to the class this way. I've never seen anyone else write react code like this.

Reply:
Ok I pushed it out to a separate function component in the same file.  Still have to have a wrapper for it in this class to push through some props.